### PR TITLE
Windows tests

### PR DESCRIPTION
**Pull request by dougthor42**
Add Windows testing via AppVeyor.

This is a prerequisite for getting proper Windows support (at least for colors - completion is another issue).

The basic idea is that paths on Windows are stupid and because we need to support Python < 3.6, we can't easily use `pathlib`. A lot of the tests are regex comparisons, and what happens is that the forward-slashed Windows paths end up looking like they have escape characters in them (namely `\p` in `...\pdbpp...`).

So we simply modify those path strings and replace all backslashes `\` with forward slashes `/`.

In addition, Windows is a case-insensitive file system, so `C:\foo` is the same as `c:\foo`. This again causes the regex-based tests to fail, so we add in `os.path.normcase` which, on windows, changes things to lowercase.

Sometimes we can't adjust the expected output of a file path (like `stacktrace.format_exc`), so in those cases we just skip checking those lines if we're testing on Windows.

Lastly there are a couple tests that I just couldn't figure out why they won't work on Windows. I figured that since they're tested on Linux, it's likely OK to skip them on Windows:

+ `test_continue_arg_with_error`: 'charmap' codec can't decode byte `0x81`.
+ `test_complete_removes_duplicates_with_coloring`: Fails on AppVeyor, but doesn't fail when I run it locally... /shrug.
+ `test_stdout_reconfigured`: I think this might be an issue with `pyreadline`. Full stack trace is:
    ```
    C:\dev_python\pdbpp\testing\test_pdb.py:5102: in test_stdout_reconfigured
        """)
    C:\dev_python\pdbpp\testing\test_pdb.py:264: in check
        expected, lines = run_func(func, expected)
    C:\dev_python\pdbpp\testing\test_pdb.py:246: in run_func
        return expected, runpdb(func, commands)
    C:\dev_python\pdbpp\testing\test_pdb.py:180: in runpdb
        func()
    C:\dev_python\pdbpp\testing\test_pdb.py:5086: in fn
        assert pdbpp.local.GLOBAL_PDB.stdout is patched_stdout
    C:\dev_python\pdbpp\testing\test_pdb.py:5086: in fn
        assert pdbpp.local.GLOBAL_PDB.stdout is patched_stdout
    C:\WinPython36_x64_zero\python-3.6.3.amd64\Lib\bdb.py:48: in trace_dispatch
        return self.dispatch_line(frame)
    C:\WinPython36_x64_zero\python-3.6.3.amd64\Lib\bdb.py:66: in dispatch_line
        self.user_line(frame)
    C:\WinPython36_x64_zero\python-3.6.3.amd64\Lib\pdb.py:261: in user_line
        self.interaction(frame, None)
    c:\dev_python\pdbpp\.tox\py36\lib\site-packages\pdbpp.py:401: in interaction
        return self._interaction(frame, traceback)
    c:\dev_python\pdbpp\.tox\py36\lib\site-packages\pdbpp.py:429: in _interaction
        self._cmdloop()
    C:\WinPython36_x64_zero\python-3.6.3.amd64\Lib\pdb.py:321: in _cmdloop
        self.cmdloop()
    C:\WinPython36_x64_zero\python-3.6.3.amd64\Lib\cmd.py:111: in cmdloop
        readline.parse_and_bind(self.completekey+": complete")
    c:\dev_python\pdbpp\.tox\py36\lib\site-packages\pyreadline\rlmain.py:100: in parse_and_bind
        self.mode._bind_key(key, func)
    c:\dev_python\pdbpp\.tox\py36\lib\site-packages\pyreadline\modes\basemode.py:165: in _bind_key
        keyinfo = make_KeyPress_from_keydescr(key.lower()).tuple()
    c:\dev_python\pdbpp\.tox\py36\lib\site-packages\pyreadline\keysyms\common.py:121: in make_KeyPress_from_keydescr
        raise IndexError("Not a valid key: '%s'"%keydescr)
    E   IndexError: Not a valid key: 'ck'
    ```

Related to #334.

### CODE REVIEW

File: src/pdbpp.py [hunk 1]
```diff
@@ -68,6 +68,37 @@ def __getattr__(name):
     raise AttributeError
 
 
+def _normalize_path(filename):
```
Comment (blueyed) [inline src/pdbpp.py]: I think this should be done only with the tests in general.

See https://github.com/pdbpp/pdbpp/pull/272/files#diff-6244362351d7c228d95cde366267b10aR39-R46 for a regexp based approach, but we could also adjust anything matching a path in the check function.

Since you have `_normpath_and_escape` already, we should do it there I guess.

**Comment by dougthor42** [comment]
Can you figure out why 5fbe71b would cause `test_question_mark_unit` to fail?

**Comment by dougthor42** [comment]
BTW, the passing AppVeyor build can be found here: https://ci.appveyor.com/project/dougthor42/pdbpp/builds/28895023

**Comment by blueyed** [comment]
Added to AppVeyor: https://ci.appveyor.com/project/blueyed/pdbpp - hopefully it will pick up the next build.

File: appveyor.yml [hunk 2]
```diff
@@ -0,0 +1,11 @@
+environment:
+  matrix:
+    - TOXENV: py36
```
Comment (blueyed) [inline appveyor.yml]: Please use "…-coverage".
Can we have py38 maybe instead already?

**Comment by blueyed** [comment]
> Can you figure out why [5fbe71b](https://github.com/pdbpp/pdbpp/commit/5fbe71b9c561b31c16422eebae03e407c8364a6e) would cause `test_question_mark_unit` to fail?

It's only comments?

File: appveyor.yml (hunk 2 above)
Comment (dougthor42) [reply appveyor.yml]: Will do. I was figuring that only the Linux builds needed coverage. Do you want the full tox suite py27,34,..,38 and pypy? Or is just running a 38 sufficient?

File: src/pdbpp.py (hunk 1 above)
Comment (dougthor42) [reply src/pdbpp.py]: Ah, yes, I'd forgotten that you already solved that issue.

I've pushed some commits that (a) reverts any changes to the core `pdbpp.py` and (b) uses the regex-style fixes that you implemented.

**Comment by dougthor42** [comment]
> It's only comments?

Yeah, that's the strange thing. No other changes

File: appveyor.yml (hunk 2 above)
Comment (dougthor42) [reply appveyor.yml]: Hmm... looks like the way you have coverage set up won't play nicely with AppVeyor. Do we need to have coverage for windows builds?

Comment (blueyed) [reply appveyor.yml]: I've just thought it would be good to have the Windows code being covered, too.
A single build should suffice.

What's the problem with AV?

**Comment by blueyed** [comment]
> > It's only comments?
> 
> Yeah, that's the strange thing. No other changes

Ok, not clear to me: do you have build log from before and after?
But don't worry, might be something flaky, and you could try by just reverting the revert maybe.

File: appveyor.yml (hunk 2 above)
Comment (blueyed) [reply appveyor.yml]: I see:
```
py38-coverage run-test: commands[1] | 'C:\projects\pdbpp/.ci/report-coverage.sh'
ERROR: InvocationError for command 'C:\projects\pdbpp/.ci/report-coverage.sh' (exited with code 8)
```

(via https://ci.appveyor.com/project/blueyed/pdbpp/builds/28897748)

Not much information there.
Would be great if you could test / fix it from trying it locally.

Comment (blueyed) [reply appveyor.yml]: You can already add `set -x` there maybe, which is good in general, and might show what fails.

Comment (dougthor42) [reply appveyor.yml]: In short, your tox env is trying to run a shell script (`coverage-report.sh`) in a Windows (CMD/PowerShell) environment.

Now, AppVeyor does support running bash scripts ([source "Run Coverage Analysis on AppVeyor"](https://www.appveyor.com/blog/2017/03/17/codecov/)):
```
after_test:
  - ps: |
      $env:PATH = 'C:\msys64\usr\bin;' + $env:PATH
      Invoke-WebRequest -Uri 'https://codecov.io/bash' -OutFile codecov.sh
      bash codecov.sh -f "coverage.xml" -t YourUploadToken
```

but I don't know how to integrate that with tox.

**Comment by dougthor42** [comment]
> 5fbe71b breaks `test_question_mark_unit`

I reverted the commit and it resolved things. In addition, with the changes made to how tests on windows are done (regex method), we no longer need the "SKIPLINEWIN" flag so the whole bit ends up not being relevant.

----

I've added some additional comments. If you're happy with this I'll rebase into logical commits and move on to the colorama stuff.

**Comment by dougthor42** [comment]
I spoke too soon... Adding comments broke `test_question_mark_unit` again. I'm investigating. It seems that it's an issue with the line numbers.

**Comment by dougthor42** [comment]
I've solved the `test_question_mark_unit` issue. Do you want 171d677 in a separate MR since it's unrelated to this?

The regex in this test was causing issues if too many lines were added
before the function definition. The test itself introspects the test_pdbpp.py
source code file and adds line numbers.

I would add comments to unrelated code near the top of the module and then
this test would break. However, if I removed N lines of comments or added M
lines, then the test would pass again.

It turns out that the issue was caused by right-aligning line numbers. At the
time of this commit, the test spanned lines 975 to 1018 after the unrelated
comments were added. This caused the LineMatcher lines to look like so
(with "|" denoting the start of line):

```
  | 998 blah
  | 999 blah
  |1000 blah
```

However, the regex was only looking for `d+` at the start of the string. The
extra whitespace from right-aligning the line numbers caused the test to break.

If fewer lines existed above this function definition, then the LineMatcher
lines would look like:

```
  |979 blah
  |980 blah
  |981 blah
```

And there would be no issue. The same applies if many more lines existed above
this function:

```
  |1003 blah
  |1004 blah
  |1005 blah
```

File: testing/test_pdb.py [hunk 3]
```diff
@@ -29,11 +29,19 @@
 pytest_plugins = ["pytester"]
 
 # Windows support
+# The basic idea is that paths on Windows are dumb because of forward slashes.
```
Comment (blueyed) [inline testing/test_pdb.py]: s/forward slashes/backslashes/?

**Comment by dougthor42** [comment]
Haha yup, I'll get that fixed.

**Comment by blueyed** [comment]
> Do you want [171d677](https://github.com/pdbpp/pdbpp/commit/171d6775a5d2b69b37002edf83f9ce09a3300d79) in a separate MR since it's unrelated to this?

Cherry-picked in 1f6571a - sorry for the trouble due to this.. !

File: appveyor.yml [hunk 4]
```diff
@@ -0,0 +1,11 @@
...
+  matrix:
+    - TOXENV: py38
+
+build: off
+
+install:
+  - pip install tox==3.13.2
+
+test_script:
+  - tox --force-dep="pytest @ git+https://github.com/dougthor42/pytest.git@win"
```
Comment (blueyed) [inline appveyor.yml]: Can use `blueyed/master` again.

File: appveyor.yml [hunk 5]
```diff
@@ -0,0 +1,11 @@
+environment:
+  matrix:
+    - TOXENV: py38
+
+build: off
+
+install:
+  - pip install tox==3.13.2
```
Comment (blueyed) [inline appveyor.yml]: Please use `python -m pip install --user git+https://github.com/blueyed/tox@master` here.

File: testing/test_pdb.py [hunk 6]
```diff
@@ -3210,12 +3224,19 @@ def fn():
...
 
 @pytest.mark.skipif(not hasattr(pdbpp.pdb.Pdb, "error"),
                     reason="no error method")
+@pytest.mark.skipif(
+    sys.platform == "win32",
+    reason=(
+        "UnicodeDecodeError: 'charmap' codec can't decode byte 0x81 in"
+        " position 6998: character maps to <undefined>"
+    ),
+)
```
Comment (blueyed) [inline testing/test_pdb.py:3264]: Would be great to figure that one out - where/how does it happen exactly?

File: testing/test_pdb.py [hunk 7]
```diff
@@ -3726,6 +3747,10 @@ def check_completions():
 """)
 
 
+@pytest.mark.skipif(
+    os.getenv("APPVEYOR", False),
+    reason="Fails in appveyor but not local Windows. Can't figure out why."
+)
```
Comment (blueyed) [inline testing/test_pdb.py]: Both variants, or only the "pyrepl" one?
Can you provide more info, e.g. the full output?

File: testing/test_pdb.py [hunk 8]
```diff
@@ -5017,6 +5042,13 @@ def test_truncate_to_visible_length(s, maxlength, expected):
 
 
 @pytest.mark.parametrize("pass_stdout", (True, False))
+@pytest.mark.skipif(
+    sys.platform == "win32",
+    reason=(
+        "IndexError: not a valid key: 'ck'. I think this is an issue"
+        " with pyreadline?"
+    ),
+)
```
Comment (blueyed) [inline testing/test_pdb.py]: It uses `completekey="ck"`, which is wrong, but nobody complained.  Should use `tab` (the default) there then.

File: appveyor.yml (hunk 2 above)
Comment (blueyed) [reply appveyor.yml]: Tox should take care of that (via https://github.com/pdbpp/pdbpp/blob/1f6571ad8dcb0d1bb5e4987996f6c4a3d1676a9d/tox.ini#L19).
Maybe it needs other slashes?

Or we just need the above env (PATH) adjustment for AppVeyor in general?

File: appveyor.yml (hunk 4 above)
Comment (dougthor42) [reply appveyor.yml]: done.

File: appveyor.yml (hunk 5 above)
Comment (dougthor42) [reply appveyor.yml]: Sure, though I am curious as to why. Why `python -m`? Why `--user`? What have you customized in your fork of tox?

File: testing/test_pdb.py (hunk 3 above)
Comment (dougthor42) [reply testing/test_pdb.py]: Silly me. Fixed

File: testing/test_pdb.py (hunk 7 above)
Comment (dougthor42) [reply testing/test_pdb.py]: Hmm... well it was both variants, but I think that may have just been an issue with my implementation. It appears to be working again, so I'll remove the skipif.

File: appveyor.yml (hunk 4 above)
Comment (dougthor42) [reply appveyor.yml]: Wait, do you mean `@my-master`?

File: appveyor.yml (hunk 5 above)
Comment (blueyed) [reply appveyor.yml]: Could be still `pip`, but `python -m pip` often helps to use the right one.
As for `--user` that might be only useful on Travis / not even work here without the user location in PATH.
There are not many customizations, but e.g. it supports `--durations` (which you could add to the call then if you like, otherwise I'd might do it in general), the intitial motivation for having the fork.
As with the pytest one I'm gonna use it for "my" projects.  It would be nice if tox would have accepted the `--durations` flag (https://github.com/tox-dev/tox/issues/1450, https://github.com/tox-dev/tox/pull/1457), but well.

File: appveyor.yml (hunk 4 above)
Comment (blueyed) [reply appveyor.yml]: Oh, sure.
Though I've thought about using "master" instead, and probably will - but then I'll fix it in one go.

File: testing/test_pdb.py (hunk 6 above)
Comment (dougthor42) [reply testing/test_pdb.py:3264]: For reference, the command I'm running for tests is:

```
tox -e py36 --force-dep="pytest @ git+https://github.com/blueyed/pytest@my-master" -- -k test_continue_arg_with_error
```

I only have Py36 available on my windows machine.

----

The full traceback is:

```pytb
C:\WinPython36_x64_zero\python-3.6.3.amd64\Lib\pdb.py:648: in do_break
    lineno = int(arg)
E   ValueError: invalid literal for int() with base 10: '.foo'

During handling of the above exception, another exception occurred:
C:\WinPython36_x64_zero\python-3.6.3.amd64\Lib\pdb.py:659: in do_break
    code = func.__code__
E   AttributeError: 'str' object has no attribute '__code__'

During handling of the above exception, another exception occurred:
C:\dev_python\pdbpp\testing\test_pdb.py:3267: in test_continue_arg_with_error
    filename=RE_THIS_FILE_CANONICAL,
C:\dev_python\pdbpp\testing\test_pdb.py:277: in check
    expected, lines = run_func(func, expected)
C:\dev_python\pdbpp\testing\test_pdb.py:259: in run_func
    return expected, runpdb(func, commands)
C:\dev_python\pdbpp\testing\test_pdb.py:193: in runpdb
    func()
C:\dev_python\pdbpp\testing\test_pdb.py:3244: in fn
    x = 1
C:\dev_python\pdbpp\testing\test_pdb.py:3244: in fn
    x = 1
C:\WinPython36_x64_zero\python-3.6.3.amd64\Lib\bdb.py:48: in trace_dispatch
    return self.dispatch_line(frame)
C:\WinPython36_x64_zero\python-3.6.3.amd64\Lib\bdb.py:66: in dispatch_line
    self.user_line(frame)
C:\WinPython36_x64_zero\python-3.6.3.amd64\Lib\pdb.py:261: in user_line
    self.interaction(frame, None)
c:\dev_python\pdbpp\.tox\py36\lib\site-packages\pdbpp.py:381: in interaction
    return self._interaction(frame, traceback)
c:\dev_python\pdbpp\.tox\py36\lib\site-packages\pdbpp.py:409: in _interaction
    self._cmdloop()
C:\WinPython36_x64_zero\python-3.6.3.amd64\Lib\pdb.py:321: in _cmdloop
    self.cmdloop()
C:\WinPython36_x64_zero\python-3.6.3.amd64\Lib\cmd.py:138: in cmdloop
    stop = self.onecmd(line)
C:\WinPython36_x64_zero\python-3.6.3.amd64\Lib\pdb.py:418: in onecmd
    return cmd.Cmd.onecmd(self, line)
C:\WinPython36_x64_zero\python-3.6.3.amd64\Lib\cmd.py:217: in onecmd
    return func(arg)
c:\dev_python\pdbpp\.tox\py36\lib\site-packages\pdbpp.py:1210: in do_continue
    self.do_tbreak(arg)
C:\WinPython36_x64_zero\python-3.6.3.amd64\Lib\pdb.py:706: in do_tbreak
    self.do_break(arg, 1)
C:\WinPython36_x64_zero\python-3.6.3.amd64\Lib\pdb.py:667: in do_break
    (ok, filename, ln) = self.lineinfo(arg)
C:\WinPython36_x64_zero\python-3.6.3.amd64\Lib\pdb.py:740: in lineinfo
    answer = find_function(item, fname)
C:\WinPython36_x64_zero\python-3.6.3.amd64\Lib\pdb.py:100: in find_function
    for lineno, line in enumerate(fp, start=1):
c:\dev_python\pdbpp\.tox\py36\lib\encodings\cp1252.py:23: in decode
    return codecs.charmap_decode(input,self.errors,decoding_table)[0]
E   UnicodeDecodeError: 'charmap' codec can't decode byte 0x81 in position 6738: character maps to <undefined>
```

Removing the `set_trace()` line from the dummy function causes the test to not error (but obviously fail):

```diff
def test_continue_arg_with_error():
     def fn():
-        set_trace()
         x = 1
         y = 2
```

Alternatively, removing the `# c.foo` line from the expected regex

```diff
@@ -3252,7 +3253,6 @@ def test_continue_arg_with_error():
 [NUM] > .*fn()
 -> x = 1
    5 frames hidden .*
-# c.foo
 \*\*\* The specified object '.foo' is not a function or was not found along sys.path.
 # c {break_lnum}
 Breakpoint NUM at {filename}:{break_lnum}
```

 will also cause the test to not error (but again, obviously fail because now there's an extra line):

```
----------------------------------------------------------------------------------------------------------- Captured stdout call -----------------------------------------------------------------------------------------------------------

\[ *[0-9]+\] > .*fn\(\)                                                                |  [36] > c:\dev_python\pdbpp\testing\test_pdb.py(3244)fn()
-> x = 1                                                                               |  -> x = 1
   5 frames hidden .*                                                                  |     5 frames hidden (try 'help hidden_frames')
\*\*\* The specified object '.foo' is not a function or was not found along sys.path.  |  # c 3246    <<<<<
# c 3246                                                                               |  Breakpoint 1 at c:\dev_python\pdbpp\testing\test_pdb.py:3246    <<<<<
Breakpoint  *[0-9]+ at c\:\\dev_python\\pdbpp\\testing\\test_pdb\.py:3246              |  Deleted breakpoint 1 at c:\dev_python\pdbpp\testing\test_pdb.py:3246    <<<<<
Deleted breakpoint  *[0-9]+                                                            |  [36] > c:\dev_python\pdbpp\testing\test_pdb.py(3246)fn()    <<<<<
\[ *[0-9]+\] > .*fn\(\)                                                                |  -> z = 3    <<<<<
-> z = 3                                                                               |     5 frames hidden (try 'help hidden_frames')    <<<<<
   5 frames hidden .*                                                                  |  # c    <<<<<
# c                                                                                    |  <None>    <<<<<
```

Removing both these lines from the expected regex causes the test to pass, but then the test is the same as `test_continue_arg`

```diff
@@ -3252,8 +3253,6 @@ def test_continue_arg_with_error():
 [NUM] > .*fn()
 -> x = 1
    5 frames hidden .*
-# c.foo
-\*\*\* The specified object '.foo' is not a function or was not found along sys.path.
 # c {break_lnum}
 Breakpoint NUM at {filename}:{break_lnum}
 Deleted breakpoint NUM
```

File: testing/test_pdb.py (hunk 8 above)
Comment (dougthor42) [reply testing/test_pdb.py]: Changing it to `tab` appears to have fixed things.

File: testing/test_pdb.py (hunk 6 above)
Comment (blueyed) [reply testing/test_pdb.py:3264]: This goes through this Python code: https://github.com/python/cpython/blob/1c2e81ed00477255b09a9c3693ddc441979c6c2e/Lib/pdb.py#L668-L676

Adding `print(arg, sys.exc_info())` there shows this in the test for me:
```
# c.foo
.foo (<class 'AttributeError'>, AttributeError("'str' object has no attribute '__code__'"), …
```
`find_function` gets called for me with `'foo' '/home/user/src/pdbpp/testing/test_pdb.py'`.

The code there is the same in 3.6.0: https://github.com/python/cpython/blob/5c4568a05a0a62b5947c55f68f9f2ecfb90a4f12/Lib/pdb.py#L92-L103

So I guess this also fails for you already?
```python
with open("testing/test_pdb.py") as fp:
    for lineno, line in enumerate(fp, start=1):
        print(lineno, repr(line))
```

Using `iconv --from=cp1252 testing/test_pdb.py` chokes at: https://github.com/pdbpp/pdbpp/blob/aabb10b07ac7db1aa5a503cddac9d08dbe58a087/testing/test_pdb.py#L3009

> def test_utf8():
    def fn():
        # Ñ‚ÐµÑiconv: illegal input sequence at position 64082

So for some reason this file is being handled as cp1252, although it is utf8, and has "# -*- coding: utf-8 -*-".

It's an interesting issue, which we should protect against in terms of not crashing in/via https://github.com/dougthor42/pdbpp/blob/aabb10b07ac7db1aa5a503cddac9d08dbe58a087/src/pdbpp.py#L1210.
From a quick glance however it appears to also be a bug in pdb itself in the first place.

I think it's ok to skip the test for now, but mark it with xfail then: https://github.com/pytest-dev/pytest/blob/63c9ad02f4b15d981adf3ec6d34560e94faf0d05/doc/en/skipping.rst#reason-parameter

Comment (dougthor42) [reply testing/test_pdb.py:3264]: K, I'll xfail it. Discussion moved to #341.

Comment (dougthor42) [reply testing/test_pdb.py:3264]: I also made the xfail very strict:
+ if a different error is raised, the test suite will fail.
+ if the test passes, the test suite will fail.

This will make it obvious when a code change affects this test.

Both only apply to Windows, of course.

Let me know if you don't want it this way.

Comment (blueyed) [reply testing/test_pdb.py:3264]: > K, I'll xfail it. Discussion moved to #341.

Awesome issue (details etc), thanks! :)
Strictness is also cool!

**Review by blueyed — COMMENTED** [review]
I assume this is ready then from your side?
Will check again later then.
Regarding squashing etc, we can just squash-merge via the UI here, if that's ok for you?

**Comment by dougthor42** [comment]
Yeah I think everything on my end is done, and `merge --squash` is fine with me. Do you need a rebase first?

I'll get started on the next part of Windows support tomorrow or this weekend.

File: appveyor.yml (hunk 2 above)
Comment (blueyed) [reply appveyor.yml]: @dougthor42 
Any more insights here?  Have you tried it locally in Windows?
I've pushed some commits to investigate / work on that, since I think we really should have coverage with it.

File: testing/test_pdb.py [hunk 9]
```diff
@@ -318,8 +342,11 @@ class Config(DefaultConfig):
         pygments.formatters.TerminalTrueColorFormatter
     )
 
-    assert Pdb(Config=Config).format_source("print(42)") == (
-        "\x1b[38;2;0;128;0;01mprint\x1b[39;00m(\x1b[38;2;102;102;102m42\x1b[39m)\n"
+    formatted_source = Pdb(Config=Config).format_source("print(42)")
+    assert formatted_source in (
+        '\x1b[38;2;0;128;0;01mprint\x1b[39;00m(\x1b[38;2;102;102;102m42\x1b[39m)\n',
+        # Might happen on AppVeyor?
+        '\x1b[38;2;0;128;0mprint\x1b[39m(\x1b[38;2;102;102;102m42\x1b[39m)\n',
```
Comment (blueyed) [inline testing/test_pdb.py]: @dougthor42 
Any idea about this?  It started already when I've started messing with it, so might be due to external deps?

**Comment by blueyed** [comment]
@dougthor42 
I've wrestled it down.. \o/
Let me know what you think - there's some XXX comment at least about some oddity with line endings, but do not think it's worth investigating (currently).

**Comment by dougthor42** [comment]
Just FYI: I haven't forgotten about this! Just been super busy 😞

**Comment by blueyed** [comment]
@dougthor42 
Thanks for letting me know.
No worries, take all the time you/it needs - I do not have much time for this myself currently also (and before when reviewing it).  Gladly we have no stale-bots here.. ;)

**Comment by blueyed** [comment]
JFI: rebased.

File: testing/test_pdb.py (hunk 9 above)
Comment (dougthor42) [reply testing/test_pdb.py]: ## TL;DR

I have no idea. I cannot get anything to return the value `\x1b[38;2;0;128;0;01m` (which includes the `;01`).

For me on both Windows and Linux (WSL):

```python
# This assert fails
assert formatted_source == '\x1b[38;2;0;128;0;01mprint\x1b[39;00m(\x1b[38;2;102;102;102m42\x1b[39m)\n'

# This assert passes
assert formatted_source == '\x1b[38;2;0;128;0mprint\x1b[39m(\x1b[38;2;102;102;102m42\x1b[39m)\n'
```


## Questions

### Question 1:

Do you know what the last `;01` in the first escape sequence is supposed to be? It looks like it's the [24-bit sequence](https://en.wikipedia.org/wiki/ANSI_escape_code#24-bit), but the Wikipedia paged doesn't list a 6th term.

```
24-bit
String:     \x1b [     38    ;  2 ;  0 ;  128 ;   0 ;  01  m
Translated:  ESC [ foreground; dim; red; green; blue; ???  m
                                                       ^
```


### Question 2:

Is there a particular reason why you wanted to change this test to use the full string? It looks like the previous iteration would pass both.

This is mostly curiosity. Using the full string is a more precise test (but more fragile as we see here).


### Question 3:

What do you get when you run this locally?

```python
>>> from pygments.formatters import TerminalTrueColorFormatter
>>> from pygments.lexers import PythonLexer
>>> from pygments import highlight
>>> highlight("print(42)", PythonLexer(), TerminalTrueColorFormatter())
```

----

The rest of this post is just me going through ideas. It can probably be ignored.

## Investigations

### The Terminals

For Windows, I have CMD and PowerShell terminals available. For Linux, I have whatever WSL uses. It looks and acts like CMD, so I assume it's the same thing behind the scenes.

I checked all 3 possible combinations and

| Terminal | OS | without `;01` in the expected string | with `;01` in the expected string |
| --- | --- | :-: | :-: |
| CMD | Windows | Pass | Fail |
| CMD | Linux (WSL) | Pass | Fail |
| PowerShell | Windows | Pass | Fail |

If I just use pygments to format that string:

```python
>>> # Python 3.6.8 on win32+CMD, pygments==2.5.2
>>> from pygments.formatters import TerminalTrueColorFormatter
>>> from pygments.lexers import PythonLexer
>>> from pygments import highlight
>>> highlight("print(42)", PythonLexer(), TerminalTrueColorFormatter())
'\x1b[38;2;0;128;0mprint\x1b[39m(\x1b[38;2;102;102;102m42\x1b[39m)\n'

>>> # Python 3.6.8 on linux+CMD, pygments==2.5.2
>>> highlight("print(42)", PythonLexer(), TerminalTrueColorFormatter())
'\x1b[38;2;0;128;0mprint\x1b[39m(\x1b[38;2;102;102;102m42\x1b[39m)\n'

>>> # Python 3.6.8 on win32+PowerShell, pygments==2.5.2
>>> highlight("print(42)", PythonLexer(), TerminalTrueColorFormatter())
'\x1b[38;2;0;128;0mprint\x1b[39m(\x1b[38;2;102;102;102m42\x1b[39m)\n'
```

In short, I have no idea what's going on.


### The Lexers

Turns out this wasn't the root cause, so this can be ignored.

I thought perhaps that the tokens for that sting were different on linux and Windows. This is not the case (they're the same), but I did notice that things change with pygments versions. Specifically, the `print` string is a `Token.Keyword` in pygments 2.4.2 and is a `Token.Name.Builtin` in pygments 2.5.2.

The tokens generated by the string `print(42)` are:

```python
>>> # Python 3.6.8 on Linux, pygments==2.4.2
>>> from pygments.lexers import PythonLexer
>>> list(PythonLexer().get_tokens("print(42)"))
[(Token.Keyword, 'print'),
 (Token.Punctuation, '('),
 (Token.Literal.Number.Integer, '42'),
 (Token.Punctuation, ')'),
 (Token.Text, '\n')]
```

```python
>>> # Python 3.6.3 on win32, pygments==2.5.2
>>> from pygments.lexers import PythonLexer
>>> list(PythonLexer().get_tokens("print(42)"))
[(Token.Name.Builtin, 'print'),
 (Token.Punctuation, '('),
 (Token.Literal.Number.Integer, '42'),
 (Token.Punctuation, ')'),
 (Token.Text, '\n')]
```

The reason this happens is because in pygments 2.4.2, the `PythonLexer` was for Python2 and a specific `Python3Lexer` existed. In pygments > 2.5, these were switched. `PythonLexer` is for Python3 and a specific `Python2Lexer` exists.


### packages

I added `python -m pip list` to the tox commands and then ran the failing test on both Linux and Windows.

On Windows, the following packages are added:
+ atomicwrites (from pytest)
+ colorama (from pytest)
+ pyreadline (from pdbpp)

It's possible, but unlikely, that one of these is mucking things up. I updated the tox env to remove `colorama` and `pyreadline`, but that had no effect. I could not remove `atomicwrites` without breaking pytest, so that wasn't checked.

I don't think this is the issue.


### Something specific in pdbpp

L733 `format_source` is what actually adds the escape codes. It determines the highlighting function and calls that as the return value.

If we force that to use the pygments lexers and formatters directly rather than going through `_get_source_highlight_function`, the I still do not see the `;01` in the output.

```diff
         src = self.try_to_decode(src)
-        return self._highlight(src)
+        import pygments
+        return pygments.highlight(src,
+            pygments.lexers.PythonLexer(stripnl=False),
+            pygments.formatters.TerminalTrueColorFormatter()
+        )
```

Basically all this does is confirm that `_get_source_highlight_function` and `_get_pygments_formatter` doesn't have any strangeness in it.

Comment (blueyed) [reply testing/test_pdb.py]: > I have no idea. I cannot get anything to return the value `\x1b[38;2;0;128;0;01m` (which includes the `;01`).

The ";01" means bold here and IIRC comes from a newer version of pygments, but have not really investigated myself.

See 3265207 (on master), where I've made the test less strict.

Comment (dougthor42) [reply testing/test_pdb.py]: > ";01" means bold here

I guess I figured that, but then second-guessed myself because the wiki page for 24-bit color doesn't mention using other modifiers.

At least for my terminal, there's no difference between with and without `;01`:

image


Ah well, no biggie. I'm fine with using the less-strict test.

**Comment by dougthor42** [comment]
Is there anything else that needs to be done on this? Once this is merged I can clean up #342.

**Comment by blueyed** [comment]
Thanks a lot! \o/
